### PR TITLE
Slackbot for Alerts

### PR DESCRIPTION
**Pull request by kavinvalli**
- Created Slackbot integration for alerts
- Each organisation can now connect to Slack as an integration
    - The "integration page" update for slack is still pending - this will be done along with the page to production work
- When the alert is being created / edited, the organisation can choose if they want to send a message in a slack channel as well - if yes
    - if slack is not connected, it shows a "Connect Slack" button
    - is slack is connected, it shows the list of channels
- After a slack channel is selected, the functionality is the same as how it was with email based alerts

### CODE REVIEW

File: valhalla/jawn/src/managers/IntegrationManager.ts [hunk 1]
```diff
@@ -94,4 +94,76 @@ export class IntegrationManager extends BaseManager {
...
+      .eq("organization_id", this.authParams.organizationId)
+      .eq("integration_name", integrationType)
+      .single();
+
+    if (error) {
+      return err(error.message);
+    }
+
+    return ok(
+      data as {
```
Comment (chitalian) [inline valhalla/jawn/src/managers/IntegrationManager.ts:121]: We are going to start using the `satisfies` keyword whereever it makes sense. It is slightly more type-safe.
```suggestion
      data satistfies {
```

File: valhalla/jawn/src/managers/IntegrationManager.ts [hunk 2]
```diff
@@ -94,4 +94,76 @@ export class IntegrationManager extends BaseManager {
...
+        id: string;
+        integration_name: string;
+        settings: Record<string, any>;
+        active: boolean;
+      }
+    );
+  }
+
+  public async getSlackChannels(): Promise<
```
Comment (chitalian) [inline valhalla/jawn/src/managers/IntegrationManager.ts:130]: very solid method!

1. you check for org id
2. you wrapped Slack in a try catch

File: web/components/templates/alerts/alertForm.tsx [hunk 3]
```diff
@@ -37,16 +44,32 @@ interface AlertFormProps {
 
 const AlertForm = (props: AlertFormProps) => {
   const { handleSubmit, onCancel, initialValues } = props;
+  const [slackRedirectUrl, setSlackRedirectUrl] = useState<string | null>(null);
+
+  useEffect(() => {
+    if (window) {
+      setSlackRedirectUrl(`${window.location.origin}/slack/redirect`);
+    }
+  }, []);
```
Comment (chitalian) [inline web/components/templates/alerts/alertForm.tsx:53]: This pattern can be reduced to a `useMemo`
```suggestion
  const slackRedirectUrl = useMemo(() => {
    if (window) {
      return`${window.location.origin}/slack/redirect`;
    }
    return null;
  }, []);
```

HOWEVER, using window.location.orgin can be drilling in from the static props that get passed from the parent component:

[see this stack overflow](https://stackoverflow.com/questions/65199051/how-to-get-page-url-or-hostname-in-nextjs-project)

File: web/pages/slack/redirect.tsx
Comment (chitalian) [inline web/pages/slack/redirect.tsx:1]: This is elite

File: web/components/templates/alerts/alertForm.tsx (hunk 3 above)
Comment (kavinvalli) [reply web/components/templates/alerts/alertForm.tsx:53]: @chitalian This is leading to passing the url through 4 components. Do we want to do that?

File: worker/src/lib/db/AlertStore.ts [hunk 4]
```diff
@@ -21,14 +26,16 @@ export class AlertStore {
   public async getAlerts(): Promise<Result<Alert[], string>> {
     const { data: alerts, error: alertsErr } = await this.supabaseClient
       .from("alert")
-      .select("*")
+      .select(
+        "*, organization (integrations (id, integration_name, settings, active))"
+      )
```
[the remark this answers is not in the document]
Comment (kavinvalli) [reply worker/src/lib/db/AlertStore.ts:31]: Yes, it is

File: valhalla/jawn/src/managers/IntegrationManager.ts (hunk 1 above)
Comment (kavinvalli) [reply valhalla/jawn/src/managers/IntegrationManager.ts]: Makes sense. Made the required changes and will keep in mind next time!

File: web/components/templates/alerts/alertForm.tsx (hunk 3 above)
Comment (chitalian) [reply web/components/templates/alerts/alertForm.tsx]: @kavinvalli good call out. it's fine - let's maybe just useMemo instead?

File: worker/src/lib/db/AlertStore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (chitalian) [reply worker/src/lib/db/AlertStore.ts:31]: @kavinvalli noice

**Review by chitalian — APPROVED** [review]

**Pull request merged**